src/library.cpp

```diff
@@ -232,14 +254,64 @@ Library::BookIdCollection Library::filter(const std::string& search)
 }
 
 
+void Library::updateBookDB(const Book& book)
+{
+  Xapian::Stem stemmer;
+  Xapian::TermGenerator indexer;
+  const std::string lang = book.getLanguage();
+  try {
+    stemmer = Xapian::Stem(iso639_3ToXapian.at(lang));
+    indexer.set_stemmer(stemmer);
+    indexer.set_stemming_strategy(Xapian::TermGenerator::STEM_SOME);
+  } catch (...) {}
+  Xapian::Document doc;
+  indexer.set_document(doc);
+
+  const std::string title = normalizeText(book.getTitle(), lang);
+  const std::string desc = normalizeText(book.getDescription(), lang);
+  doc.add_value(0, title);
+  doc.add_value(1, desc);
+  doc.set_data(book.getId());
+
+  indexer.index_text(title, 1, "S");
+  indexer.index_text(desc, 1, "XD");
+
+  // Index fields without prefixes for general search
+  indexer.index_text(title);
+  indexer.increase_termpos();
+  indexer.index_text(desc);
+
+  const std::string idterm = "Q" + book.getId();
+  doc.add_boolean_term(idterm);
+  m_bookDB.replace_document(idterm, doc);
+}
+
 Library::BookIdCollection Library::getBooksByTitleOrDescription(const Filter& filter)
 {
+  if ( !filter.hasQuery() )
+    return getBooksIds();
+
   BookIdCollection bookIds;
-  for(auto& pair:m_books) {
-    if(filter.acceptByQueryOnly(pair.second)) {
-      bookIds.push_back(pair.first);
-    }
+  Xapian::QueryParser queryParser;
+  queryParser.set_default_op(Xapian::Query::OP_AND);
+  queryParser.add_prefix("title", "S");
+  queryParser.add_prefix("description", "XD");
+  // Language assumed for the query is not known for sure so stemming
+  // is not applied
+  //queryParser.set_stemmer(Xapian::Stem(iso639_3ToXapian.at(???)));
+  //queryParser.set_stemming_strategy(Xapian::QueryParser::STEM_SOME);
+  const auto flags = Xapian::QueryParser::FLAG_PHRASE
+                   | Xapian::QueryParser::FLAG_BOOLEAN
+                   | Xapian::QueryParser::FLAG_LOVEHATE
+                   | Xapian::QueryParser::FLAG_WILDCARD;
```

Comment:
It only makes sense if the search endpoint is going to be used for suggestions. @kelson42 What's your opinion on that?

Reply:
This is used in kiwix-desktop to search in the library (local or remote). Search updated "in live" (no need to validate the search).
It seems legit that if the user search for "wiki" he gets result for all ("wikipedia", "wikispecies", ...)
If we don't, he will have no result until he enters a full word.

Reply:
I am reluctant to use the partial search mode in the backend since there is no indication of where the editing/cursor location in the text is. Though typically text is typed left to right (at least in the languages I am concerned with), it is also possible that the cursor is moved to another location and the text is edited in the middle. It is the front end (which is responsible for the editing of the text search string) that can add a hint about the cursor location in the form of a wildcard.

Reply:
> I am reluctant to use the partial search mode in the backend 
>  It is the front end [...] than can add a hint [...] in form of a wildcard

I globally agree with that.

However, as the documentation of FLAG_PARTIAL said, the wildcard should be added only in some situation (not a phrase, no a binary search, not ...). It could be pretty difficult to detect in the frontend (and we don't want to have all frontend implement it).

Maybe we could add a "boolean flag" in the api telling if we want a partial search or not and let xapian add it (if it is needed).